wcfsetup/install/files/lib/acp/form/PackageUpdateServerAddForm.class.php

```diff
@@ -31,153 +41,167 @@ class PackageUpdateServerAddForm extends AbstractForm
+                                    'invalid',
+                                    'wcf.acp.updateServer.serverURL.error.invalid',
+                                ));
+                            }
+
+                            if (
+                                ($duplicate = $this->getDuplicateServer((string)$url))
+                                && ($this->formAction === IFormDocument::FORM_MODE_CREATE
+                                || ($this->formObject instanceof PackageUpdateServer
+                                && $this->formObject->getObjectID() !== $duplicate->getObjectID()))
+                            ) {
+                                $formField->addValidationError(new FormFieldValidationError(
+                                    'duplicate',
+                                    'wcf.acp.updateServer.serverURL.error.invalid',
+                                    [
+                                        'duplicate' => $duplicate,
+                                    ]
+                                ));
+                            }
+                        })),
+                    TextFormField::create('loginUsername')
+                        ->label('wcf.acp.updateServer.loginUsername')
+                        ->description('wcf.acp.updateServer.loginUsername.description')
+                        ->maximumLength(255),
+                    PasswordFormField::create('loginPassword')
+                        ->label('wcf.acp.updateServer.loginPassword')
+                        ->description('wcf.acp.updateServer.loginPassword.description')
+                        ->placeholder($passwordPlaceholder)
+                        ->required()
+                        ->maximumLength(255)
+                        ->addDependency(
+                            NonEmptyFormFieldDependency::create('loginUsername')
+                                ->fieldId('loginUsername')
+                        ),
+                ]),
+        ]);
     }
 
     /**
      * @inheritDoc
      */
-    public function validate()
+    public function buildForm()
     {
-        parent::validate();
+        parent::buildForm();
 
-        $this->validateServerURL();
-    }
-
-    /**
-     * Validates the server URL.
-     *
-     * @since       5.3
-     */
-    protected function validateServerURL()
-    {
-        if (empty($this->serverURL)) {
-            throw new UserInputException('serverURL');
-        }
+        $this->form->getDataHandler()->addProcessor(new CustomFormDataProcessor(
+            'password',
+            function (IFormDocument $document, array $parameters) {
+                if ($this->formAction === IFormDocument::FORM_MODE_CREATE) {
+                    return $parameters;
+                }
 
-        try {
-            $url = new Uri($this->serverURL);
-            $this->serverURL = (string)$url;
+                $username = $parameters['data']['loginUsername'] ?? null;
```

Comment:
Is `null` the correct fallback value?

Reply:
I thought about adding a dependency to the username-field, so there is no need for a fallback since the index should be set in every case.